sky/skylet/providers/lambda_labs/lambda_utils.py

```diff
@@ -0,0 +1,158 @@
+"""Lambda Labs helper functions."""
+import os
+import json
+import requests
+
+CREDENTIALS_PATH = '~/.lambda_labs/lambda_keys'
+API_ENDPOINT = 'https://cloud.lambdalabs.com/api/v1'
+
+
+class LambdaLabsError(Exception):
+    __module__ = 'builtins'
+
+
+class Metadata:
+    """Local metadata for a Lambda Labs instance."""
+
+    def __init__(self, path_prefix, cluster_name):
+        # per cluster metadata file to avoid basic race conditions
+        self._metadata_path = os.path.expanduser(
+                f'{path_prefix}-{cluster_name}')
+        self._metadata = {}
+        if os.path.exists(self._metadata_path):
+            with open(self._metadata_path, 'r') as f:
+                self._metadata = json.load(f)
+
+    def __getitem__(self, instance_id):
+        return self._metadata.get(instance_id)
+
+    def __setitem__(self, instance_id, value):
+        if value is None:
+            if instance_id in self._metadata:
+                self._metadata.pop(instance_id) # del entry
+            if len(self._metadata) == 0 and \
+                    os.path.exists(self._metadata_path):
+                os.remove(self._metadata_path)
+                return
+        else:
+            self._metadata[instance_id] = value
+        with open(self._metadata_path, 'w') as f:
+            json.dump(self._metadata, f)
```

Comment:
This `Metadata` seems not to be thread-safe, as the local metadata file for a cluster will suffer from race conditions when multiple processes are operating on the same cluster.
It might be fine for now, as the SkyPilot will adopt a per-cluster lock for each ray-related operation at the higher level, but it would be good to have a comment here for a future note.

Reply:
Good idea. I will add a comment and think more about this synchronization

Reply:
I added a comment here, as you suggested!